Source/WebCore/page/PageDebuggable.cpp

```diff
@@ -34,67 +34,104 @@
 #include "Page.h"
 #include "Settings.h"
 #include <JavaScriptCore/InspectorAgentBase.h>
+#include <wtf/MainThread.h>
 #include <wtf/TZoneMallocInlines.h>
 
+
 namespace WebCore {
 
 using namespace Inspector;
 
 WTF_MAKE_TZONE_ALLOCATED_IMPL(PageDebuggable);
+Ref<PageDebuggable> PageDebuggable::create(Page& page)
+{
+    return adoptRef(*new PageDebuggable(page));
+}
 
 PageDebuggable::PageDebuggable(Page& page)
-    : m_page(page)
+    : m_page(&page)
 {
 }
 
 String PageDebuggable::name() const
 {
-    RefPtr localMainFrame = dynamicDowncast<LocalFrame>(m_page.mainFrame());
-    if (!localMainFrame)
-        return String();
+    String name;
+    callOnMainThreadAndWait([this, protectedThis = Ref { *this }, &name] {
```

Comment:
can you explain a bit of why this is necessary too?  i can understand making `m_target` guarded by a `Lock` and even the changes to `JSGlobalObjectDebuggable` just in case someone access a `JSContext` from a thread other than where it was created, but this shouldn't be reachable by anyone outside of WebKit and should only ever be created on the main thread (since it lives in the WebProcess).  what's going on for this?

Reply:
This is a cherry-pick from the security branch. I wrote the fix but do not remember exactly all the details. I'm not a WebInspector expert so if you feel like some changes in the patch are not needed, I encourage you to follow-up (likely adding assertions to guarantee this is not called off the main thread).

What I remember is that we had crashes because some of the functions was called off the main thread and the code wasn't thread safe. That said, it is entirely possible I updated code that didn't technically need to be by trying to fix surrounding code.

Reply:
can you provide any more info like repro steps or backtraces?  my understanding of how all this works is that this isn't necessary, but i also could be wrong and would love to learn if that's the case :)

Reply:
We don't have reproduction steps and the crash in the wild looked like:
```
3 __cxa_pure_virtual
                  5 Inspector::RemoteConnectionToTarget::close()::$_0::operator()() const
            ==> 5 WTF::Detail::CallableWrapper<Inspector::RemoteConnectionToTarget::close()::$_0, void>::call() <==
                      5 WTF::Function<void ()>::operator()() const
                        5 Inspector::RemoteTargetHandleRunSourceGlobal(void*)
                          5 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__
```

Reply:
hmm do you know if this was from a `WKWebView` or a `JSContext`?

Reply:
It was crashing in the SafariViewService process, it's all I know. As far as I know, the crash did go away with this fix.